### PR TITLE
WiFi pass management + usage data to ubidots

### PR DESCRIPTION
**Pull request by ihristova11**
- WiFi password management (using [WiFiManager](https://github.com/tzapu/WiFiManager))
 - Sending usage data to ubidots 
 - Email notifications when unsuccessful login attemp occurs
 
Board: 
image

Email template:
image

### CODE REVIEW

File: Config.h [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#define TOPIC "/v1.6/devices/lock"
+#define VARIABLE_LABEL_UNSUCCESSFUL "unsuccessful-attempts"
+#define VARIABLE_LABEL_SUCCESSFUL "successful-attempts"
+
+#define MQTT_SERVER "industrial.api.ubidots.com"
+#define MQTT_PORT 1883
+#define USER_TOKEN ""
+#define CLIENT_ID "Pokemon" // can be anything
```
Comment (yalishanda42) [inline Config.h:8]: pika pika

File: smart-locc.ino [hunk 2]
```diff
@@ -37,22 +49,45 @@ void setup() {
...
 
         // TODO: update lock state, display info, etc
         // ...
         //
-
+        if (currentState == DeviceState::AUTHORIZED || currentState == DeviceState::AUTH_FAILURE)
+        {
+            char data[250];
+            sprintf(data, "{\"%s\":{\"value\":1}}",
+                    currentState == DeviceState::AUTHORIZED ? VARIABLE_LABEL_UNSUCCESSFUL : VARIABLE_LABEL_UNSUCCESSFUL);
```
Comment (yalishanda42) [inline smart-locc.ino]: One of those should be `VARIABLE_LABEL_SUCCESSFUL` I guess

File: smart-locc.ino [hunk 3]
```diff
@@ -37,22 +49,45 @@ void setup() {
...
         Serial.print("STATE -> ");
         Serial.println(currentState);
 
         // TODO: update lock state, display info, etc
         // ...
         //
-
+        if (currentState == DeviceState::AUTHORIZED || currentState == DeviceState::AUTH_FAILURE)
+        {
+            char data[250];
```
Comment (yalishanda42) [inline smart-locc.ino]: 250 is a lot of bytes, better write it like
```cpp
String data = currentState == DeviceState::AUTHORIZED ? VARIABLE_LABEL_SUCCESSFUL : VARIABLE_LABEL_UNSUCCESSFUL
```

Comment (allexks) [reply smart-locc.ino:81]: Or alternatively, do not declare those 2 constants using `#define` but instead have them be `const char*` in `Config.h`

File: smart-locc.ino [hunk 4]
```diff
@@ -37,22 +49,45 @@ void setup() {
...
+            sprintf(data, "{\"%s\":{\"value\":1}}",
+                    currentState == DeviceState::AUTHORIZED ? VARIABLE_LABEL_UNSUCCESSFUL : VARIABLE_LABEL_UNSUCCESSFUL);
+            client.publish(TOPIC, data);
+        }
     }
 
     button.update();
-    if (button.pressed() && currentState == DeviceState::IDLE) {
+    if (button.pressed() && currentState == DeviceState::IDLE)
+    {
```
Comment (yalishanda42) [inline smart-locc.ino:86]: ;(

**Review by yalishanda42 — CHANGES_REQUESTED** [review]

File: smart-locc.ino [hunk 5]
```diff
@@ -37,22 +49,45 @@ void setup() {
...
+        Serial.print("Reconnecting to server...");
+        bool connected = client.connect(CLIENT_ID, USER_TOKEN, PASS);
+        Serial.println(connected ? "success" : "failure");
+    }
+    else
+    {
+        client.loop();
+    }
+    delay(2000);
```
Comment (yalishanda42) [inline smart-locc.ino]: This is happening on absolutely every iteration, I don't think we want that lag

**Review by yalishanda42 — CHANGES_REQUESTED** [review]

File: smart-locc.ino (hunk 5 above)
Comment (ihristova11) [reply smart-locc.ino]: Done

File: smart-locc.ino (hunk 3 above)
Comment (ihristova11) [reply smart-locc.ino]: Updated

File: smart-locc.ino (hunk 2 above)
Comment (ihristova11) [reply smart-locc.ino]: The truth has been spoken :D

**Review by yalishanda42 — APPROVED** [review]

**Pull request merged**